### PR TITLE
Add cache to PageFunctionCompiler

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

**Comment by nezihyigitbasi** [comment]
Compilation fails in travis jobs.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 1]
```diff
@@ -99,15 +105,84 @@
...
+    private final LoadingCache<RowExpression, Supplier<PageFilter>> filterCache;
+
+    private final CacheStatsMBean projectionCacheStats;
+    private final CacheStatsMBean filterCacheStats;
+
     @Inject
-    public PageFunctionCompiler(Metadata metadata)
+    public PageFunctionCompiler(Metadata metadata, CompilerConfig config)
+    {
+        this(metadata, config.isExpressionCacheEnabled());
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: Just to be consistent with similar classes add `requireNonNull` for the args.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestPageProcessorCompiler.java [hunk 2]
```diff
@@ -78,7 +79,8 @@ public void testNoCaching()
     public void testSanityRLE()
             throws Exception
     {
-        PageProcessor processor = new ExpressionCompiler(createTestMetadataManager())
+        MetadataManager metadata = createTestMetadataManager();
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestPageProcessorCompiler.java]: I think we can get rid of the repetition in these tests if we do:

```
    @BeforeMethod
    public void setup()
    {
        metadataManager =  createTestMetadataManager();
        compiler = new ExpressionCompiler(metadataManager, new PageFunctionCompiler(metadataManager, false));
    }

    @AfterMethod(alwaysRun = true)
    public void  tearDown()
    {
        metadataManager = null;
        compiler = null;
    }
```

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestPageProcessorCompiler.java [hunk 3]
```diff
@@ -62,7 +63,7 @@ public void testNoCaching()
             throws Throwable
     {
         MetadataManager metadata = METADATA_MANAGER;
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestPageProcessorCompiler.java]: only one test is using this static `METADATA_MANAGER`, to be consistent with the other tests we can remove this and create a new instance with before/after methods (see below).

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestCompilerConfig.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import static io.airlift.configuration.testing.ConfigAssertions.assertFullMapping;
+import static io.airlift.configuration.testing.ConfigAssertions.assertRecordedDefaults;
+import static io.airlift.configuration.testing.ConfigAssertions.recordDefaults;
+
+public class TestCompilerConfig
+{
+    @Test
+    public void testDefaults()
+    {
+        assertRecordedDefaults(recordDefaults(CompilerConfig.class)
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestCompilerConfig.java]: I though we should do:

```
assertRecordedDefaults(
            recordDefaults(CompilerConfig.class)
            .setInterpreterEnabled(false)
            .setExpressionCacheEnabled(true));
```

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 5]
```diff
@@ -99,15 +105,84 @@
...
+                    });
+            projectionCacheStats = new CacheStatsMBean(projectionCache);
+        }
+        else {
+            projectionCache = null;
+            projectionCacheStats = null;
+        }
+
+        if (expressionCacheEnabled) {
+            filterCache = CacheBuilder.newBuilder().recordStats().maximumSize(10_000).build(
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: Should the `10_000` be configurable too?

File: presto-main/src/main/java/com/facebook/presto/testing/LocalQueryRunner.java [hunk 6]
```diff
@@ -290,7 +292,8 @@ private LocalQueryRunner(Session defaultSession, FeaturesConfig featuresConfig,
         this.accessControl = new TestingAccessControlManager(transactionManager);
         this.pageSourceManager = new PageSourceManager();
 
-        this.expressionCompiler = new ExpressionCompiler(metadata);
+        this.pageFunctionCompiler = new PageFunctionCompiler(metadata, false);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/testing/LocalQueryRunner.java]: in fact, why not `true` here?

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestPageFunctionCompiler.java [hunk 7]
```diff
@@ -80,7 +80,7 @@ public void testFailureDoesNotCorruptFutureResults()
     @Test
     public void testGeneratedClassName()
     {
-        PageFunctionCompiler functionCompiler = new PageFunctionCompiler(createTestMetadataManager());
+        PageFunctionCompiler functionCompiler = new PageFunctionCompiler(createTestMetadataManager(), false);
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestPageFunctionCompiler.java]: Consider adding a test here exercising  caching `PageFunctionCompiler`.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 8]
```diff
@@ -99,15 +105,84 @@
...
+    {
+        return filterCacheStats;
     }
 
     public Supplier<PageProjection> compileProjection(RowExpression projection, Optional<String> classNameSuffix)
     {
+        if (projectionCache == null) {
+            return compileProjectionInternal(projection, classNameSuffix);
+        }
+        return projectionCache.getUnchecked(projection);
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: The suffix is a hint to the compiler.  We recently added this so when the cache is disabled you get the queryId and plan node in the class name, and thus, the stack traces.

Also we are not sure the cache is even needed.  This came up from a very small benchmark sf100 on a 100 node cluster, but reasonable sized workloads don't need a cache.  The plan is to add this and then test in some real world workloads.

File: presto-main/src/main/java/com/facebook/presto/testing/LocalQueryRunner.java (hunk 6 above)
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/testing/LocalQueryRunner.java]: The default is true so it is tested by all of the integration tests.  I set the rest of the tests to false to test the disabled case.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestCompilerConfig.java (hunk 4 above)
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestCompilerConfig.java]: That would result in an indentation like this:
```
assertRecordedDefaults(
            recordDefaults(CompilerConfig.class)
                    .setInterpreterEnabled(false)
                    .setExpressionCacheEnabled(true));

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 9]
```diff
@@ -99,15 +106,86 @@
...
+                    });
+            projectionCacheStats = new CacheStatsMBean(projectionCache);
+        }
+        else {
+            projectionCache = null;
+            projectionCacheStats = null;
+        }
+
+        if (expressionCacheSize > 0) {
+            filterCache = CacheBuilder.newBuilder().recordStats().maximumSize(expressionCacheSize).build(
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: Same as above

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 10]
```diff
@@ -99,15 +106,86 @@
...
+    @Managed
+    @Nested
+    public CacheStatsMBean getFilterCache()
+    {
+        return filterCacheStats;
     }
 
     public Supplier<PageProjection> compileProjection(RowExpression projection, Optional<String> classNameSuffix)
     {
+        if (projectionCache == null) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: I tried it, but this version seemed more readable.

**Pull request closed without merging**